### PR TITLE
Add names of "regular" cluster features to extracted feature metadata

### PR DESCRIPTION
**Pull request by ldematte**
Today we extract all historical features to a JSON map that is parsed by ESRestTestCase to provide info about historical features. 
This PR adds to that map the list of all names (Ids) of regular features too, so we can reliably raise an error if a feature is not defined.

The logic we have today to detect if a feature is "off" or "non-existing" (wrong, non-existing id) is flawed and works only for historical features. This fixes it. The distinction works only for non-legacy tests; for legacy tests, we log a warning and skip the check.

### CODE REVIEW

**Comment by ldematte** [comment]
@elasticmachine update branch

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 1]
```diff
@@ -2345,32 +2355,61 @@ private static class ESRestTestCaseHistoricalFeatures implements FeatureSpecific
...
-
-                historicalFeatures = Collections.unmodifiableMap(historicalFeaturesMap);
             }
+            return historicalFeaturesMap;
+        }
 
-            return historicalFeatures;
+        static Set<String> loadFeatureNames(String metadataPath) {
+            Set<String> featureNames = new HashSet<>();
+            String[] metadataFiles = metadataPath.split(System.getProperty("path.separator"));
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:2396]: Can use `File.separator` instead

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 2]
```diff
@@ -2345,32 +2355,61 @@ private static class ESRestTestCaseHistoricalFeatures implements FeatureSpecific
...
+            for (String metadataFile : metadataFiles) {
+                try (
+                    InputStream in = Files.newInputStream(PathUtils.get(metadataFile));
+                    XContentParser parser = JsonXContent.jsonXContent.createParser(XContentParserConfiguration.EMPTY, in)
+                ) {
+                    Map<String, Object> parsedMap = parser.map();
+                    if (parsedMap.get("feature_names") instanceof ArrayList<?> parsedFeatureNames) {
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:2403]: Shouldn't be checking against concrete types - should be `Collection<?>` instead

File: test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java [hunk 3]
```diff
@@ -8,8 +8,30 @@
...
+import java.util.function.Predicate;
+
+public interface TestFeatureService {
     TestFeatureService ALL_FEATURES = ignored -> true;
+
+    static Predicate<String> createHistoricalFeaturePredicate(
+        NavigableMap<Version, Set<String>> historicalFeatures,
+        Collection<Version> nodeVersions
+    ) {
+        var minNodeVersion = nodeVersions.stream().min(Version::compareTo);
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java:25]: Use `Comparator.naturalOrder()` again

File: test/metadata-extractor/src/test/java/org/elasticsearch/extractor/features/HistoricalFeaturesMetadataExtractorTests.java [hunk 4]
```diff
@@ -33,15 +40,30 @@ public class HistoricalFeaturesMetadataExtractorTests extends ESTestCase {
...
+            assertThat(parsedMap, hasKey("historical_features"));
+            assertThat(parsedMap, hasKey("feature_names"));
+            @SuppressWarnings("unchecked")
+            Map<String, Object> historicalFeaturesMap = (Map<String, Object>) (parsedMap.get("historical_features"));
             for (Map.Entry<NodeFeature, Version> entry : nodeFeatureVersionMap.entrySet()) {
-                assertThat(parsedMap, hasEntry(entry.getKey().id(), entry.getValue().toString()));
+                assertThat(historicalFeaturesMap, hasEntry(entry.getKey().id(), entry.getValue().toString()));
+            }
+            @SuppressWarnings("unchecked")
+            ArrayList<String> featureNamesList = (ArrayList<String>) (parsedMap.get("feature_names"));
```
Comment (thecoop) [inline test/metadata-extractor/src/test/java/org/elasticsearch/extractor/features/HistoricalFeaturesMetadataExtractorTests.java:64]: Using concrete type again

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 5]
```diff
@@ -2345,32 +2355,61 @@ private static class ESRestTestCaseHistoricalFeatures implements FeatureSpecific
...
+                            historicalFeaturesMap.put(
+                                new NodeFeature((String) entry.getKey()),
+                                Version.fromString((String) entry.getValue())
+                            );
                         }
-                    } catch (IOException e) {
-                        throw new UncheckedIOException(e);
                     }
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:2388]: Do we need to catch & hide IOException here? Can't it just propagate?

File: test/metadata-extractor/src/main/java/org/elasticsearch/extractor/features/HistoricalFeaturesMetadataExtractor.java [hunk 6]
```diff
@@ -62,23 +65,36 @@ public void generateMetadataFile(Path outputFile) {
...
+                generator.writeEndArray();
+            });
             generator.writeEndObject();
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
-    public Map<NodeFeature, Version> extractHistoricalFeatureMetadata() {
+    void extractHistoricalFeatureMetadata(CheckedBiConsumer<Map<NodeFeature, Version>, Set<String>, IOException> metadataConsumer)
```
Comment (thecoop) [inline test/metadata-extractor/src/main/java/org/elasticsearch/extractor/features/HistoricalFeaturesMetadataExtractor.java:88]: Why a biconsumer here? Why not return `Tuple<Map<,>, Set<>>`?

Comment (ldematte) [reply test/metadata-extractor/src/main/java/org/elasticsearch/extractor/features/HistoricalFeaturesMetadataExtractor.java:88]: I don't like tuples, they are too "opaque" - they make it difficult to tell what is what and to see types

**Comment by ldematte** [comment]
After discussion offline with @thecoop , we decided this is not probably worth it. I'll close this PR for now and submit another one with a simple fix to the wrong assertion.

**Pull request closed without merging**